dbms/src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -717,19 +720,29 @@ void InterpreterSelectQuery::executeImpl(Pipeline & pipeline, const BlockInputSt
 }
 
 
-static void getLimitLengthAndOffset(ASTSelectQuery & query, size_t & length, size_t & offset)
+void InterpreterSelectQuery::getLimitLengthAndOffset(ASTSelectQuery & query, size_t & length, size_t & offset)
 {
     length = 0;
     offset = 0;
     if (query.limit_length)
     {
-        length = safeGet<UInt64>(typeid_cast<ASTLiteral &>(*query.limit_length).value);
+        getLimitUIntValue(query.limit_length, length);
         if (query.limit_offset)
-            offset = safeGet<UInt64>(typeid_cast<ASTLiteral &>(*query.limit_offset).value);
+            getLimitUIntValue(query.limit_offset, offset);
     }
 }
 
 
+void InterpreterSelectQuery::getLimitUIntValue(const ASTPtr& ptr, size_t& result)
+{
+    const auto& eval_result = evaluateConstantExpression(ptr, context);
+    if (!isNumber(eval_result.second)) {
+        throw Exception("Illegal limit expression", ErrorCodes::INVALID_LIMIT_EXPRESSION);
+    }
+    result = applyVisitor(FieldVisitorConvertToNumber<UInt64>(), eval_result.first);
```

Comment:
Overflow may happen. Expressions like `LIMIT -1` must be explicitly disabled.

Reply:
The simplest way is to check field type explicitly and process different cases for UInt64, Int64, Float64.
For Float64 we should also check that the limit is a whole number.